source/opt/propagator.h

```diff
@@ -0,0 +1,242 @@
+
+  // Returns true if |instr| should be simulated again. This tests whether the
+  // instruction is in the set of instructions not to simulate again.
+  bool ShouldSimulateAgain(ir::Instruction* instr) const {
+    return do_not_simulate_.find(instr) != do_not_simulate_.end();
+  }
+
+  // Add |instr| to the set of instructions not to simulate again.
+  void DontSimulateAgain(ir::Instruction* instr) {
+    do_not_simulate_.insert(instr);
+  }
+
+  // Returns true if |block| has been simulated already.
+  bool IsBlockSimulated(ir::BasicBlock* block) {
+    return simulated_blocks_.find(block) != simulated_blocks_.end();
+  }
+
+  // Marks block |block| as simulated.
+  void MarkBlockSimulated(ir::BasicBlock* block) {
+    simulated_blocks_.insert(block);
+  }
+
+  // Marks |edge| as executable.  Returns false if the edge was already marked
+  // as executable. Otherwise, the edge is marked executable for the first time
+  // and it returns true.
+  bool MarkEdgeExecutable(const Edge& edge) {
+    return executable_edges_.insert(edge).second;
+  }
+
+  // Returns true if |edge| has been marked as executable.
+  bool IsEdgeExecutable(const Edge& edge) {
+    return executable_edges_.find(edge) != executable_edges_.end();
+  }
+
+  // Return a pointer to the def-use manager for this context.
+  analysis::DefUseManager* get_def_use_mgr() const {
+    return ctx_->get_def_use_mgr();
+  }
+
+  // Add the destination block of the CFG edge |e| to the work list.  If |e|
+  // had already been executed, do nothing.
+  void AddControlEdge(const Edge& e);
+
+  // Add all the use statements at the end of SSA edges coming out the
+  // definition for |id|.
+  void AddSSAEdges(uint32_t id);
+
+  // Function that visit instructions during simulation. The output of this
+  // function is used to determine if the simulated instruction produced a value
+  // interesting for propagation. The actual value produced by the instruction
+  // must be stored in the |values_| map.
+  VisitFunction visit_fn_;
+
+  // IR context to use.
+  ir::IRContext* ctx_;
+
+  // CFG to propagate values on. TODO(dnovillo): The CFG should be part of
+  // IRContext.
+  ir::CFG* cfg_;
+
+  // SSA def-use edges to traverse. Each entry is a destination statement for an
+  // SSA def-use edge as returned by |def_use_manager_|.
+  std::queue<ir::Instruction*> ssa_edges_;
+
+  // Blocks to simulate.
+  std::queue<ir::BasicBlock*> blocks_;
+
+  // Blocks simulated during propagation.
+  std::unordered_set<ir::BasicBlock*> simulated_blocks_;
+
+  // CFG blocks in structured order. Every block in this list comes after its
+  // dominators.
+  std::list<ir::BasicBlock*> ordered_blocks_;
+
+  // Set of instructions that should not be simulated again because they have
+  // been found to be in the Varying state.
+  std::unordered_set<ir::Instruction*> do_not_simulate_;
+
+  // Map between a basic block and its predecessor edges.
+  std::unordered_map<ir::BasicBlock*, std::list<Edge>> bb_preds_;
```

Comment:
Please repeat the todo about moving these into CFG (which is to be moved into IRContext).

Reply:
Done.